commons-core/src/main/scala/com/avsystem/commons/serialization/macroCodecs.scala

```diff
@@ -143,38 +143,43 @@ abstract class FlatSealedHierarchyCodec[T](
   final def readObject(input: ObjectInput): T = {
     val oooFields = new FieldValues(oooFieldNames, oooDeps, typeRepr)
 
-    def read(): T =
-      if (input.hasNext) {
-        val fi = input.nextField()
-        if (fi.fieldName == caseFieldName) {
-          val caseName = readCaseName(fi)
-          caseIndexByName(caseName) match {
-            case -1 => unknownCase(caseName)
-            case idx => readFlatCase(caseName, oooFields, input, caseDeps(idx))
-          }
-        } else if (!oooFields.tryReadField(fi)) {
-          if (caseDependentFieldNames.contains(fi.fieldName)) {
-            if (defaultCaseIdx != -1) {
-              val defaultCaseName = caseNames(defaultCaseIdx)
-              val wrappedInput = new DefaultCaseObjectInput(fi, input, defaultCaseName)
-              readFlatCase(defaultCaseName, oooFields, wrappedInput, caseDeps(defaultCaseIdx))
-            } else {
-              missingCase(fi.fieldName)
-            }
+    def readCase(caseNameField: FieldInput): T = {
+      val caseName = readCaseName(caseNameField)
+      caseIndexByName(caseName) match {
+        case -1 => unknownCase(caseName)
```

Comment:
These barbaric checks could be avoided if `caseIndexByName` would just return `Opt[GenCodec[_: T]`

Reply:
Yeah but I can't similarly change `caseIndexByValue` because in `writeObject` I need to access both `caseNames` and `caseDeps` arrays. So I have to work with indices anyway.